File: chatgpt/gpt2table.pl
#!/usr/bin/perl

#$infile = "./ChatGPT_genefunction.md";
#$outfile = "./ChatGPT_genefunction_table.tsv";

$infile = "./ChatGPT_genefunction2b.md";
$outfile = "./ChatGPT_genefunction_table2b.tsv";

open(IN,"$infile");
@lines = <IN>;
close(IN);
$nlines = @lines;

open(OUT,">$outfile");
print OUT "Source\tGeneSetID\tGeneSetName\tGeneList\tn_Genes\tLLM Name\tLLM Analysis\tScore\n";
for($i = 0;$i < $nlines;$i++) {
	$l = $lines[$i];
	chomp $l;
	if ($l =~ /^\*\*Process\:.+\(([\d\.]+?)\)\*\*/) {
		$score = $1;
		$drid = '';
		$drid = $1 if ($prev2 =~ /\*\*(.+?)\*\*/);
	} elsif ($l =~ /^\*\*(\d+)\. (.+?)\*\*/) {
		$num = $1;
		$func = $2;
		$cluster = "Cluster\_$drid\_$num";
		$i++;
		$l2 = $lines[$i];
		undef @ita;
		$i++;
		$l3 = $lines[$i];
		chomp $l3;
		$j = $i;
		$l3t = '';
		while ($j != 0) {
#			while ($l3 =~ /[^\*]\*([^\*]+?)\*[^\*]/g) {
			while ($l3 =~ /(?<!\*)\*([^\*]+?)\*(?!\*)/g) {
				push @ita, $1;
			}
			$j2 = $j + 2;
			$l3t .= "$l3 ";
			if ($lines[$j2] !~ /^\*\*/) {
				$l3 = $lines[$j2];
				chomp $l3;
				$j = $j2;
			} else {
				$i = $j;
				$j = 0;
				$l3 = $l3t
			}
		}
		undef %gene;
		foreach $x (@ita) {
			@ita2 = split(/\, /,$x);
			foreach $y (@ita2) {
				$gene{$1} = 1 if ($y =~ /^([A-Z][A-Z0-9]+)/);
#					push @genes, $1;
			}
		}
		@genes = (sort keys %gene);
		$gn = @genes;
		$glist = join(' ',@genes);
		print OUT "NeST\t$cluster\t$cluster\t$glist\t$gn\t$func\t$l3\t$score\n";
		print "NeST\t$cluster\t$cluster\t$glist\t$gn\t$func\t$l3\t$score\n";
#		$aaa = <STDIN>;
	}
	$prev2 = $prev1;
	$prev1 = $l;
}

File: chatgpt/gpt2table.py
import re

infile = "ChatGPT_genefunction.md"
outfile = "ChatGPT_genefunction_table2.tsv"

with open(infile, "r", encoding="utf-8") as f:
    lines = f.readlines()

nlines = len(lines)
output_lines = ["Source\tGeneSetID\tGeneSetName\tGeneList\tn_Genes\tLLM Name\tLLM Analysis\tScore"]

prev2 = ""
prev1 = ""
i = 0

while i < nlines:
    l = lines[i].strip()

    # スコアとデバイスIDの取得
    match_score = re.match(r"\*\*Process\:.+\(([\d\.]+?)\)\*\*", l)
    if match_score:
        score = match_score.group(1)
        drid = ''
        match_drid = re.match(r"\*\*(.+?)\*\*", prev2)
        if match_drid:
            drid = match_drid.group(1)
        i += 1
        continue

    # 機能セクションの検出
    match_func = re.match(r"\*\*(\d+)\. (.+?)\*\*", l)
    if match_func:
        num = match_func.group(1)
        func = match_func.group(2)
        cluster = f"Cluster_{drid}_{num}"

        i += 1
        if i >= nlines:
            break
        l2 = lines[i].strip()

        i += 1
        if i >= nlines:
            break
        l3 = lines[i].strip()

        paragraph_lines = []
        while i + 2 < nlines and not lines[i + 2].startswith("**"):
            paragraph_lines.append(l3)
            i += 2
            l3 = lines[i].strip()
        paragraph_lines.append(l3)
        full_paragraph = " ".join(paragraph_lines)

        # イタリック部（遺伝子候補）の抽出
        italic_matches = re.findall(r'(?<!\*)\*([^\*]+?)\*(?!\*)', full_paragraph)
        gene_set = set()
        for match in italic_matches:
            candidates = match.split(', ')
            for y in candidates:
                gene_match = re.match(r'^([A-Z][A-Z0-9]+)', y)
                if gene_match:
                    gene_set.add(gene_match.group(1))

        genes = sorted(gene_set)
        glist = " ".join(genes)
        gn = len(genes)

        output_lines.append(f"NeST\t{cluster}\t{cluster}\t{glist}\t{gn}\t{func}\t{full_paragraph}\t{score}")

    prev2 = prev1
    prev1 = l
    i += 1

# 出力ファイルへ書き込み
with open(outfile, "w", encoding="utf-8") as f:
    for line in output_lines:
        f.write(line + "\n")

print(f"✅ 出力ファイルを作成しました: {outfile}")

